paasta_tools/utils.py

```diff
@@ -3806,3 +3807,33 @@ def _reorder_docker_volumes(volumes: List[DockerVolume]) -> List[DockerVolume]:
         v["containerPath"].rstrip("/") + v["hostPath"].rstrip("/"): v for v in volumes
     }.values()
     return sort_dicts(deduped)
+
+
+@time_cache(ttl=5000)
+def get_instance_type_to_k8s_namespace() -> Mapping[str, str]:
+    instance_type_to_k8s_namespace = {}
+
+    def default_cr_id(service: str, insatnce: str) -> Mapping[str, str]:
+        """Fallback for modules which don't have a cr_id function"""
+        return dict()
+
+    for instance_type in INSTANCE_TYPES:
+        try:
+            instance_type_module = importlib.import_module(
+                f"paasta_tools.{instance_type}_tools"
+            )
+        except ModuleNotFoundError:
+            # paasta_native defies the convention...
+            log.warning(
+                f"Cannot lookup k8s namespace no module called paasta_tools.{instance_type}_tools"
+            )
+            continue
+        # First attempt to use cr_id() to get the CR definition and namespace
+        cr_id = getattr(instance_type_module, "cr_id", default_cr_id)("", "")
+        if "namespace" in cr_id:
+            instance_type_to_k8s_namespace[instance_type] = cr_id["namespace"]
+        # Otherwise, for objects that don't have a cr_id(), fallback to `KUBERNETES_NAMESPACE`
+        instance_type_to_k8s_namespace[instance_type] = getattr(
+            instance_type_module, "KUBERNETES_NAMESPACE", "paasta"
+        )
+    return instance_type_to_k8s_namespace
```

Comment:
thoughts on having a simple dict of `{instance_type: namespace}`? 

might be a little clearer for supporting new instance types in the future - with this change, developers kinda need to either know that they have to set a KUBERNETES_NAMESPACE constant or define a `cr_id()` function whereas with a plain dict the mapping is all in one place (and they don't need to worry about what takes precedence)

it'd also make this code a little safer too - I personally always get a little worried by dynamic imports/getattrs since they circumvent mypy and are a little harder to follow as a human reading the code

(feel free to close this if you disagree though!)

Reply:
I do mostly agree with you, I'd prefer something clearer aswell (I was thinking a decorator on the `cr_id` function that adds it), but just having a single dict is even easier. The only worry I have with the dict-approach is that people will forget to add their stuff there, but since there is a fallback to `paasta` as namespace I dont think it would cause any issues.